pkg/dispatcher/dispatcher/handler.go

```diff
@@ -0,0 +1,254 @@
+		_ = h.consumeMessage(message, destinationURL, replyURL, deadLetterURL)
+
+		// Mark The Message As Having Been Consumed (Does Not Imply Successful Delivery - Only Full Retry Attempts Made)
+		session.MarkMessage(message, "")
+	}
+
+	// Return Success
+	return nil
+}
+
+// Consume A Single Message
+func (h *Handler) consumeMessage(consumerMessage *sarama.ConsumerMessage, destinationURL *url.URL, replyURL *url.URL, deadLetterURL *url.URL) error {
+
+	// Debug Log Kafka ConsumerMessage
+	h.Logger.Debug("Consuming Kafka Message",
+		zap.Any("Headers", consumerMessage.Headers),
+		zap.ByteString("Key", consumerMessage.Key),
+		zap.ByteString("Value", consumerMessage.Value),
+		zap.String("Topic", consumerMessage.Topic),
+		zap.Int32("Partition", consumerMessage.Partition),
+		zap.Int64("Offset", consumerMessage.Offset))
+
+	// Convert The Sarama ConsumerMessage Into A CloudEvents Message
+	message := kafkasaramaprotocol.NewMessageFromConsumerMessage(consumerMessage)
+	if message.ReadEncoding() == binding.EncodingUnknown {
+		h.Logger.Warn("Received A Message With Unknown Encoding - Skipping")
+		return errors.New("received a message with unknown encoding - skipping")
+	}
+
+	// Create A New Retry Runner With Configured Backoff Behavior
+	retryRunner := retry.New(
+		retry.Config{
+			DisableBackoff: !h.ExponentialBackoff, // Note Negation: Env Var Is Whether To Enable & Retry Config Is Whether To Disable ; )
+			Times:          h.calculateNumberOfRetries(),
+			WaitBase:       time.Millisecond * time.Duration(h.InitialRetryInterval),
+		},
+	)
+
+	// Attempt To Dispatch The CloudEvent Message Via Knative Message Dispatcher With Retry Wrapper
+	err := retryRunner.Run(context.Background(), func(ctx context.Context) error {
+		return h.sendMessage(ctx, message, destinationURL, replyURL)
+	})
+
+	// Retries Failed
+	if err != nil {
+
+		// If The DeadLetterSink Is Configured
+		if deadLetterURL != nil {
+
+			// Then Make One Final Attempt With It & Return Result
+			h.Logger.Info("Making One Final Attempt With DeadLetterSink Configured")
+			err := h.MessageDispatcher.DispatchMessage(context.Background(), message, nil, destinationURL, replyURL, deadLetterURL)
+			if err != nil {
+				h.Logger.Error("Failed To Send Final Attempt With DeadLetterSink", zap.Error(err))
+			}
+			return err
+
+		} else {
+
+			// Otherwise Return Original Retry Failure (DeadLetter Sink Not Configured)
+			h.Logger.Error("Failed To Send After Configured Number Of Retries - DeadLetterSink Not Configured", zap.Error(err))
+			return err
+		}
+	}
+
+	// Return Success
+	return nil
+}
+
+// Message Sending Functionality For Use Within Retry Runner (With Custom StatusCode Handling)
+func (h *Handler) sendMessage(ctx context.Context, message *kafkasaramaprotocol.Message, destinationURL *url.URL, replyURL *url.URL) error {
+
+	// Dispatch The Message
+	err := h.MessageDispatcher.DispatchMessage(ctx, message, nil, destinationURL, replyURL, nil)
```

Comment:
the message dispatcher now has a retry system built-in, you might be interested

Reply:
Ah very cool - always good to reduce this code and re-use the framework wherever possible.   I'd like to tackle that refactor as part of a separate issue to upgrade/catch-up to the latest eventing & eventing-contrib (current PR is still based on slightly older versions).  I've created #35 with a note to do that refactor - thanks!